src/main.rs

```diff
@@ -100,64 +125,98 @@ async fn announce(req: HttpRequest, data: web::Data<AppState>) -> HttpResponse {
             seed_count_mod -= 1;
             post_announce_pipeline.cmd("ZREM").arg(&seeders_key).arg(&parsed.ip_port).ignore(); // We dont care about the return value
-        } else if is_leecher {
+        } else if let Exists::Yes = is_leecher_v2 {
             leech_count_mod -= 1;
             post_announce_pipeline.cmd("ZREM").arg(&leechers_key).arg(&parsed.ip_port).ignore(); // We dont care about the return value
         }
     } else if parsed.is_seeding {
 
         // New seeder
-        if is_seeder == false {
+        if let Exists::No = is_seeder_v2 {
             post_announce_pipeline.cmd("ZADD").arg(&seeders_key).arg(time_now_ms).arg(&parsed.ip_port).ignore();
             seed_count_mod += 1
         }
 
         // They just completed
         if event == "completed" {
             // If they were previously leecher, remove from that pool
-            if is_leecher {
+            if let Exists::Yes = is_leecher_v2 {
                 post_announce_pipeline.cmd("ZREM").arg(&leechers_key).arg(&parsed.ip_port).ignore();
                 leech_count_mod -= 1
             }
 
             // Increment the downloaded count for the infohash stats
             post_announce_pipeline.cmd("HINCRBY").arg(&parsed.info_hash).arg("downloaded").arg(1u32).ignore();
         }
-    } else if is_leecher == false {
+    } else if let Exists::No = is_leecher_v2 {
             post_announce_pipeline.cmd("ZADD").arg(&leechers_key).arg(time_now_ms).arg(&parsed.ip_port).ignore();
             leech_count_mod += 1
     };
 
-    // Update seeder & leecher count, if required
-    if seed_count_mod != 0 {
-        post_announce_pipeline.cmd("HINCRBY").arg(&parsed.info_hash).arg("seeders").arg(seed_count_mod).ignore();
-    }
+    // In future, we will use this to determine if we need to update the cache or not
+    if seed_count_mod == 0 && leech_count_mod == 0 {
+        post_announce_pipeline.cmd("INCR").arg(constants::NOCHANGE_ANNOUNCE_COUNT_KEY).ignore();
+    } else {
+        // Something changed, so we should update seeders / leechers:
+        if seed_count_mod != 0 {
+            post_announce_pipeline.cmd("HINCRBY").arg(&parsed.info_hash).arg("seeders").arg(seed_count_mod).ignore();
+        }
 
-    if leech_count_mod != 0 {
-        post_announce_pipeline.cmd("HINCRBY").arg(&parsed.info_hash).arg("leechers").arg(leech_count_mod).ignore();
-    }
+        if leech_count_mod != 0 {
+            post_announce_pipeline.cmd("HINCRBY").arg(&parsed.info_hash).arg("leechers").arg(leech_count_mod).ignore();
+        }
 
-    // endex = end index XD. seems in rust cannot select first 50 elements, or limit to less if vector doesnt have 50
-    // e.g. &seeders[0..50] is panicking when seeders len is < 50. Oh well.
-    let seeder_endex = if seeders.len() >= 50 {
-        50
-    } else {
-        seeders.len()
-    };
+        // We also need to prepare a fresh body (instead of using cache):
+        refresh_flag = true;
```

Comment:
basically we need to invalidate the cache, so if we reset `cached_reply` to be empty, we can switch on `cached_reply.len()` later on